core-rust/state-manager/src/limits.rs

```diff
@@ -0,0 +1,179 @@
+ * losses of any kind, in tort, contract or otherwise (including but not limited to loss
+ * of revenue, income or profits, or loss of use or data, or loss of reputation, or loss
+ * of any economic or other opportunity of whatsoever nature or howsoever arising), arising
+ * out of or in connection with (without limitation of any use, misuse, of any ledger system
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+use node_common::config::limits::VertexLimitsConfig;
+use radix_engine::transaction::ExecutionMetrics;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum VertexLimitsExceeded {
+    TransactionsCount,
+    TransactionsSize,
+    ExecutionCostUnitsConsumed,
+    SubstateReadSize,
+    SubstateReadCount,
+    SubstateWriteSize,
+    SubstateWriteCount,
+}
+
+pub struct VertexLimitsTracker {
+    pub total_transactions_count: usize,
+    pub total_transactions_size: usize,
+    pub total_execution_cost_units_consumed: usize,
+    pub total_substate_read_size: usize,
+    pub total_substate_read_count: usize,
+    pub total_substate_write_size: usize,
+    pub total_substate_write_count: usize,
+}
+
+impl VertexLimitsTracker {
+    pub fn new() -> Self {
+        Self {
+            total_transactions_count: 0,
+            total_transactions_size: 0,
+            total_execution_cost_units_consumed: 0,
+            total_substate_read_size: 0,
+            total_substate_read_count: 0,
+            total_substate_write_size: 0,
+            total_substate_write_count: 0,
+        }
+    }
+
+    pub fn check_pre_execution(
```

Comment:
This is fine, but if we refactor, I might consider to track the remaining instead of the total, and combine the "advance" and "check_" functions to be `consume_pre_execution(..) -> Result<(), VertexLimitsExceeded>`

Reply:
> track the remaining instead of the total

I like this approach (it would also bypass the "`limits_config` passed everywhere" issue https://github.com/radixdlt/babylon-node/pull/510#discussion_r1234445596).
The only downsides are:
- errors won't surface the nice "got `X`, but max is `Y`" messages _(not a big problem for us, since our `Y` is not very dynamic)_
- some broad "constraints" cannot be tracked in a "countdown" manner _(also not yet our problem, since all our constraints so far are of type "limit")_

@malexandru-rdx (just a suggestion, you decide) I wonder if we would benefit from introducing a proper "math"-like abstractions here...
E.g. a `struct VertexSize { transactions_count, transactions_bytes, substate_reads_count, ... }`.
And then:
- the `VertexLimitsConfig` is gone (i.e. it is implemented as `pub config_max_vertex_size: VertexSize,`)
- the `VertexLimitsTracker` simply has `pub current: VertexSize, pub limit: VertexSize,`
- you can introduce "math" like `(self.current + added) < self.limit` or whatever you find helpful

Reply:
> This is fine, but if we refactor, I might consider to track the remaining instead of the total, and combine the "advance" and "check_" functions to be consume_pre_execution(..) -> Result<(), VertexLimitsExceeded>

Initially I wrote it with a single `try_advance` method, but I realised it's not ideal to execute transaction if it doesn't fit by size anyway.
I agree merging `check_post_execution` with `advance` though.

Reply:
> (just a suggestion, you decide) I wonder if we would benefit from introducing a proper "math"-like abstractions here...
Nice idea, only thing I don't like it's we still have to add 7 ifs to get the corresponding `VertexLimitExcedeed`.